core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/recovery/RecoveryQueue.scala

```diff
@@ -0,0 +1,153 @@
+import edu.uci.ics.amber.engine.architecture.worker.WorkerInternalQueue
+import edu.uci.ics.amber.engine.architecture.worker.WorkerInternalQueue.{
+  ControlElement,
+  EndMarker,
+  EndOfAllMarker,
+  InputTuple,
+  InternalQueueElement,
+  SenderChangeMarker
+}
+import edu.uci.ics.amber.engine.common.virtualidentity.ActorVirtualIdentity
+import lbmq.LinkedBlockingMultiQueue
+
+import java.util.concurrent.LinkedBlockingQueue
+import scala.annotation.tailrec
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+
+class RecoveryQueue(logReader: DeterminantLogReader) {
+  private val records = logReader.mkLogRecordIterator()
+  private val inputMapping = mutable
+    .HashMap[ActorVirtualIdentity, LinkedBlockingQueue[InternalQueueElement]]()
+  private val controlMessages = mutable
+    .HashMap[ActorVirtualIdentity, mutable.Queue[ControlElement]]()
+  private var step = 0L
+  private var targetVId: ActorVirtualIdentity = _
+  private var currentInputSender: ActorVirtualIdentity = _
+  private var cleaned = false
+  private val callbacksOnEnd = new ArrayBuffer[() => Unit]()
+  private var endCallbackTriggered = false
+  private var nextRecordToEmit: InternalQueueElement = _
+
+  // calling it first to get nextRecordToEmit ready
+  // we assume the log has the following structure:
+  // Ctrl -> [StepDelta] -> Ctrl -> [StepDelta] -> EOF|Ctrl
+  processInternalEventsTillNextControl()
+
+  def registerOnEnd(callback: () => Unit): Unit = {
+    callbacksOnEnd.append(callback)
+  }
+
+  def isReplayCompleted: Boolean = {
+    val res = !records.hasNext
+    if (res && !endCallbackTriggered) {
+      endCallbackTriggered = true
+      callbacksOnEnd.foreach(callback => callback())
+    }
+    res
+  }
+
+  def drainAllStashedElements(
+      dataQueue: LinkedBlockingMultiQueue[Int, InternalQueueElement]#SubQueue,
+      controlQueue: LinkedBlockingMultiQueue[Int, InternalQueueElement]#SubQueue
+  ): Unit = {
+    if (!cleaned) {
+      getAllStashedInputs.foreach(dataQueue.add)
+      getAllStashedControls.foreach(controlQueue.add)
+      cleaned = true
+    }
+  }
+
+  def add(elem: InternalQueueElement): Unit = {
+    elem match {
+      case tuple: InputTuple =>
+        currentInputSender = tuple.from
+        inputMapping
+          .getOrElseUpdate(tuple.from, new LinkedBlockingQueue[InternalQueueElement]())
+          .put(tuple)
+      case SenderChangeMarker(newUpstreamLink) =>
+      //ignore, we use log to enforce original order
+      case control: ControlElement =>
+        controlMessages
+          .getOrElseUpdate(control.from, new mutable.Queue[ControlElement]())
+          .enqueue(control)
+      case WorkerInternalQueue.EndMarker =>
+        inputMapping
+          .getOrElseUpdate(currentInputSender, new LinkedBlockingQueue[InternalQueueElement]())
+          .put(EndMarker)
+      case WorkerInternalQueue.EndOfAllMarker =>
```

Comment:
remove endofall maker in internal queue element, move the logic from worker actor to data processor

Reply:
done